src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/HPackDecoder.cs

```diff
@@ -235,125 +396,106 @@ private void DecodeInternal(ReadOnlySpan<byte> data, bool endHeaders, IHttpHeade
+            if (val == 0)
+            {
+                _state = State.HeaderNameLength;
+                ParseHeaderNameLength(data, ref currentIndex, handler);
+            }
+            else
+            {
+                if (_integerDecoder.BeginTryDecode((byte)val, indexPrefix, out var intResult))
+                {
+                    OnIndexedHeaderName(intResult);
+                    ParseHeaderValueLength(data, ref currentIndex, handler);
+                }
+                else
+                {
+                    _state = State.HeaderNameIndex;
+                    ParseHeaderNameIndex(data, ref currentIndex, handler);
+                }
+            }
+        }
 
-                        if (_integerDecoder.BeginTryDecode((byte)(b & ~HuffmanMask), StringLengthPrefix, out intResult))
-                        {
-                            OnStringLength(intResult, nextState: State.HeaderValue);
+        private void ParseHeaderName(ReadOnlySpan<byte> data, ref int currentIndex, IHttpHeadersHandler? handler)
+        {
+            // Read remaining chars, up to the length of the current data
+            var count = Math.Min(_stringLength - _stringIndex, data.Length - currentIndex);
 
-                            if (intResult == 0)
-                            {
-                                ProcessHeaderValue(handler);
-                            }
-                        }
-                        else
-                        {
-                            _state = State.HeaderValueLengthContinue;
-                        }
+            // Check whether the whole string is available in the data and no decompressed required.
+            // If string is good then mark its range.
+            // NOTE: it may need to be copied to buffer later the if value is not current data.
+            if (count == _stringLength && !_huffman)
+            {
+                // Fast path. Store the range rather than copying.
+                _headerNameRange = new Range(currentIndex, currentIndex + count);
+                currentIndex += count;
 
-                        break;
-                    case State.HeaderValueLengthContinue:
-                        if (_integerDecoder.TryDecode(b, out intResult))
-                        {
-                            // IntegerDecoder disallows overlong encodings where an integer is encoded with more bytes than is strictly required.
-                            // 0 should always be represented by a single byte, so we shouldn't need to check for it in the continuation case.
-                            Debug.Assert(intResult != 0, "A header value length of 0 should never be encoded with a continuation byte.");
+                _state = State.HeaderValueLength;
+            }
+            else
+            {
+                // Copy string to temporary buffer.
+                data.Slice(currentIndex, count).CopyTo(_stringOctets.AsSpan(_stringIndex));
```

Comment:
How do we know _stringOctets is large enough to store the data being copied?  That was previously guaranteed somewhere earlier?

Reply:
OnStringLength is called before parsing the string. If _stringOctets is too small then it is expanded. There is the potential for some inefficiency because it will get expanded even if it is not used, but the array is reused forever so I didn't try to optimize that scenario.